metadata-transformer/src/main/java/no/unit/nva/metadata/extractors/DateExtractor.java

```diff
@@ -0,0 +1,63 @@
+package no.unit.nva.metadata.extractors;
+
+import no.unit.nva.metadata.type.DcTerms;
+import no.unit.nva.model.EntityDescription;
+import no.unit.nva.model.PublicationDate;
+import nva.commons.core.JacocoGenerated;
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Statement;
+
+import java.util.Set;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+public final class DateExtractor {
+    public static final String DATE_SEPARATOR = "-";
+    public static final int FULL_DATE = 3;
+    public static final int YEAR_ONLY = 1;
+    public static final Set<IRI> DATE_PREDICATES = Set.of(DcTerms.DATE.getIri(), DcTerms.DATE_ACCEPTED.getIri(),
+            DcTerms.DATE_COPYRIGHTED.getIri(), DcTerms.DATE_SUBMITTED.getIri());
+    public static final String DATE_REGEX = "^[0-9]{4}(-[0-9]{2}(-[0-9]{2})?)?$";
+    public static final Function<ExtractionPair, EntityDescription> apply = DateExtractor::extract;
+
+
+    @JacocoGenerated
+    private DateExtractor() {
+
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    private static EntityDescription extract(ExtractionPair extractionPair) {
+        Statement statement = extractionPair.getStatement();
+        EntityDescription entityDescription = extractionPair.getEntityDescription();
+        if (isDate(statement)) {
+            addDate(entityDescription, statement);
+        }
+        return entityDescription;
+    }
+
+    private static boolean isDate(Statement candidate) {
+        return DATE_PREDICATES.contains(candidate.getPredicate())
+                && Pattern.matches(DATE_REGEX, candidate.getObject().stringValue());
+    }
+
+    private static void addDate(EntityDescription entityDescription, Statement statement) {
+        if (!DcTerms.DATE.getIri().equals(statement.getPredicate())) {
+            return; // We don't yet know what should be mapped
+        }
+        String date = statement.getObject().stringValue();
+        String[] dateParts = date.split(DATE_SEPARATOR);
+        int length = dateParts.length;
+
+        PublicationDate.Builder publicationDateBuilder = new PublicationDate.Builder()
+                .withYear(dateParts[0]);
+
+        if (length > YEAR_ONLY) {
+            publicationDateBuilder.withMonth(dateParts[1]);
+        }
+        if (length == FULL_DATE) {
```

Comment:
I find the use of this constant more confusing than helpful. I have to scroll up and confirm that it is a number, then scroll back down and remember the number to see it in the context of the code. Changing the name of the variable length to dateLength or dateFormatLength, or maybe introducing some kind of data format type would probably help a little too, but we might want to consider whether always making constants out of things is beneficial.

All of this is subjective, but in this specific context I personally find the dataParts indexes more "magic" than having the values in a simple int comparison there "on paper" as it where.

Reply:
Actually, if we're going to stick with constants named after the format of the date, don't call the int anything to do with length at all. At that point it should be dateFormat.